File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/features/conversation/TaskHomeDevicePolicy.kt
package com.tabtin.mobile.features.conversation

/**
 * 任务页只读设备状态条的数据策略。
 *
 * 会话只按 Workspace 过滤；这里只聚合当前可执行 Workspace 所依赖的设备。
 */
internal object TaskHomeDevicePolicy {
    private const val MAX_NAME_WIDTH: Double = 8.0
    private const val LATIN_CHAR_WIDTH: Double = 0.55

    data class DeviceItem(
        val id: String,
        val fullName: String,
        val shortName: String,
        val isOffline: Boolean,
    )

    data class DeviceInput(
        val id: String,
        val name: String?,
        val isOffline: Boolean,
    )

    fun items(
        workspaceDeviceIds: List<String?>,
        devices: List<DeviceInput>,
        fallbackName: String,
    ): List<DeviceItem> {
        val usedDeviceIds = workspaceDeviceIds.mapNotNull { it?.trim()?.takeIf(String::isNotEmpty) }.toSet()
        return devices.mapNotNull { device ->
            if (device.id !in usedDeviceIds) return@mapNotNull null
            val fullName = device.name?.trim()?.takeIf(String::isNotEmpty) ?: fallbackName
            DeviceItem(
                id = device.id,
                fullName = fullName,
                shortName = shortName(fullName),
                isOffline = device.isOffline,
            )
        }.sortedWith(compareBy({ it.isOffline }, { it.fullName }, { it.id }))
    }

    /** 单设备在线时不占一行；多设备或任一离线时才展示。 */
    fun shouldShowRail(items: List<DeviceItem>): Boolean =
        items.size > 1 || items.any { it.isOffline }

    private fun shortName(name: String): String {
        if (displayWidth(name) <= MAX_NAME_WIDTH) return name
        val budget = MAX_NAME_WIDTH - LATIN_CHAR_WIDTH
        var used = 0.0
        val kept = StringBuilder()
        for (char in name) {
            val width = charWidth(char)
            if (used + width > budget) break
            used += width
            kept.append(char)
        }
        return if (kept.isEmpty()) "…" else "$kept…"
    }

    private fun displayWidth(text: String): Double = text.sumOf(::charWidth)

    private fun charWidth(char: Char): Double =
        if (char.code >= 0x2E80) 1.0 else LATIN_CHAR_WIDTH
}

File: apps/tabtin-ios/Tabtin/Features/Home/TaskHomeDevicePolicy.swift
import Foundation

/// 任务页设备状态条的数据策略。
///
/// 会话只按 Workspace 过滤；设备状态条不可交互，只把当前组织可执行 Workspace
/// 所依赖的设备聚合成一份简短状态。没有被任何 Workspace 使用的设备不展示。
enum TaskHomeDevicePolicy {
    static let maxNameWidth: Double = 8
    private static let latinCharWidth: Double = 0.55

    struct DeviceItem: Identifiable, Equatable, Hashable {
        let id: String
        let fullName: String
        let shortName: String
        let isOffline: Bool
    }

    static func items(
        workspaceDeviceIds: [String?],
        devices: [(id: String, name: String?, isOffline: Bool)],
        fallbackName: String
    ) -> [DeviceItem] {
        let usedDeviceIds = Set(workspaceDeviceIds.compactMap(normalized))
        return devices.compactMap { device in
            guard usedDeviceIds.contains(device.id) else { return nil }
            let fullName = normalized(device.name) ?? fallbackName
            return DeviceItem(
                id: device.id,
                fullName: fullName,
                shortName: shortName(fullName),
                isOffline: device.isOffline
            )
        }
        .sorted {
            if $0.isOffline != $1.isOffline { return !$0.isOffline }
            if $0.fullName != $1.fullName { return $0.fullName < $1.fullName }
            return $0.id < $1.id
        }
    }

    /// 单设备在线时没有需要常驻展示的信息；多设备或任一离线时才占一行。
    static func shouldShowRail(items: [DeviceItem]) -> Bool {
        items.count > 1 || items.contains { $0.isOffline }
    }

    static func shortName(_ name: String) -> String {
        guard displayWidth(name) > maxNameWidth else { return name }
        let budget = maxNameWidth - latinCharWidth
        var used: Double = 0
        var kept = ""
        for character in name {
            let width = characterWidth(character)
            guard used + width <= budget else { break }
            used += width
            kept.append(character)
        }
        return kept.isEmpty ? "…" : kept + "…"
    }

    private static func displayWidth(_ text: String) -> Double {
        text.reduce(0) { $0 + characterWidth($1) }
    }

    private static func characterWidth(_ character: Character) -> Double {
        guard let scalar = character.unicodeScalars.first else { return latinCharWidth }
        return scalar.value >= 0x2E80 ? 1 : latinCharWidth
    }

    private static func normalized(_ value: String?) -> String? {
        let trimmed = value?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return trimmed.isEmpty ? nil : trimmed
    }
}
